stream_alert/alert_processor/outputs.py

```diff
@@ -803,3 +805,296 @@ def dispatch(self, **kwargs):
+                         resp.status_code,
+                         resp.content)
+        return jira_id
+
+    def _get_comments(self, issue_id):
+        """Get all comments for an existing Jira issue
+
+        Args:
+            issue_id (str): The existing issue ID or key
+
+        Returns:
+            list: List of comments associated with a Jira issue
+        """
+        comment_url = os.path.join(self.base_url, self.COMMENT_ENDPOINT.format(issue_id))
+        resp = self._get_request(comment_url,
+                                 headers=self._get_headers(),
+                                 verify=False)
+        comments = []
+
+        try:
+            comments = resp.json()['comments']
+        except (ValueError, KeyError):
+            LOGGER.error('An error occurred while getting Jira comments [%s] : [%s]',
+                         resp.status_code,
+                         resp.content)
+        return comments
+
+    def _get_existing_issue(self, issue_summary, project_key):
+        """Find an existing Jira issue based on the issue summary
+
+        Args:
+            issue_summary (str): The Jira issue summary
+            project_key (str): The Jira project to search
+
+        Returns:
+            int: ID of the found issue or False if existing issue does not exist
+        """
+        jql = 'summary ~ "{}" and project="{}"'.format(issue_summary, project_key)
+        resp = self._search_jira(jql, fields=['id', 'summary'], max_results=1)
+        jira_id = False
+
+        try:
+            jira_id = resp[0]['id']
+        except (IndexError, KeyError):
+            LOGGER.debug('Existing Jira issue not found')
+
+        return jira_id
+
+    def _create_issue(self, issue_name, project_key, issue_type, description):
+        """Create a Jira issue to write alerts to. Alert is written to the "description"
+        field of an issue.
+
+        Args:
+            issue_name (str): The name of the Jira issue
+            project_key (str): The Jira project key which issues will be associated with
+            issue_type (str): The type of issue being created
+            description (str): The body of text which describes the issue
+
+        Returns:
+            int: ID of the created issue or False if unsuccessful
+        """
+        issue_url = os.path.join(self.base_url, self.ISSUE_ENDPOINT)
+        issue_body = {
+            'fields': {
+                'project': {
+                    'key': project_key
+                },
+                'summary': issue_name,
+                'description': description,
+                'issuetype': {
+                    'name': issue_type
+                }
+            }
+        }
+
+        resp = self._post_request(issue_url,
+                                  data=issue_body,
+                                  headers=self._get_headers(),
+                                  verify=False)
+
+        try:
```

Comment:
While I'm a fan of your pythonic approach here (using try/except), I'm realizing we should _always_ expect the response (if there is one) to be json. With that in mind, this could be consolidated to:

```
resp = self._post_...
if not self._check_http_response(resp):    # If this returns False, `_check_http_response` will log the error
    return False

response = resp.json()

return response.get('id', False)
```

Reply:
I can simplify this a bit more.  For all Jira responses, if a successful status code is passed back, I think you can make the assumption that valid JSON is also passed back, though I tend to be a little paranoid around exactly what is returned from the api calls. 

This is what the method looked like before consolidating the error conditions, when I was leveraging both _check_http_response(), as well as the try/catch to validate that the json content contained the expected data.

        resp = self._post_request(issue_url, data=issue_body, headers=headers, verify=False)
        if not self._check_http_response(resp):
            LOGGER.error('An error occurred during Jira issue creation %s', resp.content)
            return False

        try:
            issue_id = resp_dict.json()['id']
        except (ValueError, KeyError):
            LOGGER.error('An error occurred when parsing Jira issue creation %s', resp.content)
            return False

        return issue_id

One thing I just noticed when looking into this is that _check_http_response() might need to be tweaked, as it will most likely raise an exception at this [line](https://github.com/airbnb/streamalert/blob/master/stream_alert/alert_processor/output_base.py#L227).  If a non 2XX response is passed back, I don't think there is any guarantee that valid json is passed, and assuming that valid json is passed back, I think it is likely that [this](https://github.com/airbnb/streamalert/blob/master/stream_alert/alert_processor/output_base.py#L230) will raise a KeyError exception, as 'messages' and 'errors' are not guaranteed to be present.

This is what _check_http_respons() looked like back when I attempted the urllib -> requests transition, when I had a check to validate json:  

    @staticmethod
    def _check_http_response(resp, validate_json=False):
        if validate_json:
            try:
                resp.json()
            except ValueError:
                return False
        return resp and (200 <= resp.status_code <= 299)


Getting back to your original comments, let me know if you want me to modify all of the methods to remove the try/except, and if you want me to tweak _check_http_response().

Reply:
I was thinking the same thing re `_check_http_response` potentially causing an exception. I'd be good with you adding a try/except in that function similar to your example that does something like:
```
try:
    content = resp.json()
except ValueError:
    content = resp.content

# Then we could still log the content, even if it's not json
...
```
To your point, though, it wouldn't raise a KeyError because it uses `get`on the response.

Hopefully to answer you question - tweak away (_check_http_response, etc)! If you can make the change one place and benefit from it 5x places, I'm all for it.

Reply:
Ah my mistake.  Whenever I use "get" its been with a default value, so I just assumed it would also raise a KeyError exception if the key was missing, rather than returning None.

I'll update _check_http_response and clean up the try/excepts.

Reply:
Just out of curiosity, is it necessary to make the distinction between these two:

    LOGGER.error('Encountered an error while sending to %s: %s\n%s',
                  cls.__service__,
                  content.get('message'),
                  content.get('errors'))

and 

    LOGGER.error('Encountered an error while sending to %s:\n%s',
                  cls.__service__,
                  resp.content)

The later would include the "message" and "error" values in resp.content should they be present, along with everything else passed back with the response.  My preference would be to do away with the former unless you have any objections.

Reply:
Sure - that's fine, I suppose you're right that they'd all be present in the content. Ship it!